src/lib/phdi-building-blocks/phdi_building_blocks/fhir.py

```diff
@@ -125,3 +128,184 @@ def upload_bundle_to_fhir_server(bundle: dict, access_token: str, fhir_url: str)
+        raise requests.HTTPError(response=response)
+
+
+def export_from_fhir_server_poll(
+    poll_url: str, access_token: str, poll_step: float = 30, poll_timeout: float = 300
+) -> requests.Response:
+    """Poll for export file avialability after an export has been initiated.
+
+    :param poll_url: URL to poll for export information
+    :param access_token: Bearer token used for authentication
+    :param poll_step: the number of seconds to wait between poll requests, waiting
+    for export files to be generated. defaults to 30
+    :param poll_timeout: the maximum number of seconds to wait for export files to
+    be generated. defaults to 300
+    :raises polling.TimeoutException: If the FHIR server continually returns a 202
+    status indicating in progress until the timeout is reached.
+    :raises requests.HTTPError: If an unexpected status code is returned.
+    :return: The export response obtained from the FHIR server (200 status code)
+    """
+    response = polling.poll(
+        target=__export_from_fhir_server_poll_call,
+        args=[poll_url, access_token],
+        step=poll_step,
+        timeout=poll_timeout,
+    )
+
+    # Handle error conditions
+    if response.status_code != 200:
+        raise requests.HTTPError(
+            f"Encountered status {response.status_code} when requesting status"
+            + "of export `{poll_url}`"
+        )
+
+    # If no error conditions, return response
+    return response
+
+
+def download_from_export_response(
+    export_response: dict,
+) -> Iterator[Tuple[str, TextIO]]:
+    """Accepts the export response content as specified here:
+    https://hl7.org/fhir/uv/bulkdata/export/index.html#response---complete-status
+
+    Loops through the "output" array and yields the resource_type (e.g. Patient)
+    along with TextIO wrapping ndjson content.
+
+    :param export_response: export response JSON
+    :yield: tuple containing resource type (e.g. Patient) AND THE TextIO of the
+    downloaded ndjson content for that resource type
+    """
+    # TODO: Handle error array that could be contained in the response content.
+
+    for export_entry in export_response.get("output", []):
+        resource_type = export_entry.get("type")
+        blob_url = export_entry.get("url")
+        yield (resource_type, _download_export_blob(blob_url=blob_url))
+
+
+def _download_export_blob(blob_url: str, encoding: str = "utf-8") -> TextIO:
+    """Download an export file blob.
+
+    :param blob_url: Blob URL location to download from Azure Blob storage
+    :param encoding: encoding to apply to the ndjson content, defaults to "utf-8"
+    :return: Downloaded content wrapped in TextIO
+    """
+    bytes_buffer = io.BytesIO()
+    cred = DefaultAzureCredential()
+    download_blob_from_url(blob_url=blob_url, output=bytes_buffer, credential=cred)
+    text_buffer = io.TextIOWrapper(buffer=bytes_buffer, encoding=encoding, newline="\n")
+    text_buffer.seek(0)
+    return text_buffer
```

Comment:
I'm a little confused about this workflow. So the FHIR server export endpoint sticks all of the patient resources in blob storage (probably silver for us, maybe gold, whatever lol) as ndjson files. Our poller will get a 200 back when that process is complete. At that point, we start downloading the ndjson files and converting them to TextIO. It doesn't seem like we actually need to do this for exporting itself (which seems to terminate the second we get that 200 back), so is the download more about setting up future work on these files (e.g. conversion to parquet)?

Reply:
Yeah, the download doesn't happen automatically after the export.  The new function does call the 2 processes in sequence (export, download), but they are separable.